all-sdk/external-client-configuration.md

```diff
@@ -0,0 +1,340 @@
+such as samples and Terraform providers.
+
+Temporal needs a way for users to provide client configuration without hardcoding in CLI arguments or SDK options. This
+will allow code to seamlessly switch between profiles such as between a dev-server and cloud. The same configuration
+must work across all SDKs/CLIs.
+
+### Goals
+
+* Common, well-specified configuration format usable by Temporal tooling for client options
+* Implementation in all Temporal SDKs for reading and writing of this configuration
+* All Temporal CLI/SDKs updated to have easy client creation using this configuration
+* Common on-disk location in every platform for reading/writing this format
+* Ability to provide environment variables to represent values in the configuration
+* Well-specified hierarchy for on-disk overridable by environment variables overridable by in-CLI/SDK values
+* 0 new dependencies in SDKs
+* Profile names within the configuration
+
+### Non-goals
+
+* Common format for declarative workers, schedules, etc. This is only for client options.
+* File-based hierarchy. We don't need many levels of file merging for these options at this time.
+* Extreme focus on human authoring of this format. This is because it'd likely go against the "0 new dependencies" rule.
+* Use external configuration by default. Users need to opt-in to this for compatibility and clarity reasons (but it is
+  of course easy).
+
+## Specification
+
+### Values and File Format
+
+* The format is a JSON object with the following fields:
+  * `profiles` - Object with keys as profile names. The default profile should be named `default`.
+    * `<name>` - Profile object with configuration values. Profile names are case-insensitive. 💭 Why? They are needed
+      for environment variables which are case insensitive. It is a validation error to have a config file with two
+      separately-cased profile names that are equal case-insensitively.
+      * `address` - Client address, aka gRPC "host:port".
+      * `namespace` - Client namespace.
+      * `apiKey` - Client API key.
+      * `tls` - A boolean (true is same as empty object) _or_ an object. Note the default for this value is dependent
+        upon the client and may change. 💭 Why? We regret not making TLS the default in the past. If TLS is an object,
+        it can have the following possible fields:
+        * `clientCertPath` - File path to mTLS client cert. Mutually exclusive with `clientCertData`.
+        * `clientCertData` - Cert data. Mutually exclusive with `clientCertPath`.
+        * `clientKeyPath` - File path to mTLS client key. Mutually exclusive with `clientKeyData`.
+        * `clientKeyData` - Key data. Mutually exclusive with `clientKeyPath`.
+        * `serverCaCertPath` - File path to server CA cert. Mutually exclusive with `serverCaCertData`.
+        * `serverCaCertData` - CA cert data. Mutually exclusive with `serverCaCertPath`.
```

Comment:
Camel case is getting awkward and hard to predict here (people might think it's "CA"). Is there a strong reason for it? Did you consider `server-ca-cert-data`, `api-key` and `client-cert-path`?

Reply:
This is proper JSON formatting of key names, but we can use hyphens if we prefer them. They are translated to language idiomatic forms except for the CLI but yeah we don't want to have CLI translating from one to the other. I don't mind hyphen-based.

Reply:
Yes, this is proper JSON formatting of key names. And especially for our case, since it will be proto JSON, there is no real way to do hyphens well. Most users won't use the file directly, but if we are concerned with how this looks in the CLI, maybe the CLI can translate to hyphens in either direction? It is important that any user with any lang can just read this in as proto JSON and immediately have a well-typed object to mutate, and we can't get that with hyphens.

Reply:
If hyphens don't work well then let's consider underscores. This is of course subjective and up to preference, but it's a design decision we have to make. I don't usually see camel case in config files -- can you give examples?

> Yes, this is proper JSON formatting of key names.

Not sure what you meant by "proper".

Reply:
> Not sure what you meant by "proper".

JSON formatting, specifically of protobuf in our case but as a best practice for JSON in general, expects camel case key names.

Reply:
Well I was certainly about to respond with "Citation?" :) I do see that in Google's style guide and some others, which I'd forgotten. But there are multiple entire language communities that don't use camel case for anything -- there's no blanket statement to be made about JSON keys.

Reply:
Even beyond best practice, in this case I'm specifically proposing using a proto model to define the contract for the configuration, which has strict JSON expectations: https://protobuf.dev/programming-guides/proto3/#json.